packages/data-flow/src/registries/event/cachedEventRegistry.ts

```diff
@@ -0,0 +1,70 @@
+import { NewProcessedEvent, ProcessedEvent } from "@grants-stack-indexer/repository";
+import { ChainId, ILogger, stringify } from "@grants-stack-indexer/shared";
+
+import { IEventsRegistry } from "../../internal.js";
+
+/**
+ * Proxy class to cache the events in memory or fallback to another event registry
+ */
+export class InMemoryCachedEventRegistry implements IEventsRegistry {
+    private cache: Map<ChainId, ProcessedEvent> = new Map();
+
+    private constructor(
+        private readonly logger: ILogger,
+        private readonly eventRegistry: IEventsRegistry,
+        cache: Map<ChainId, ProcessedEvent>,
+    ) {
+        this.cache = structuredClone(cache);
+    }
+
+    /** @inheritdoc */
+    async getLastProcessedEvent(chainId: ChainId): Promise<ProcessedEvent | undefined> {
+        const cachedEvent = this.cache.get(chainId);
+        if (cachedEvent) {
+            return cachedEvent;
+        }
```

Comment:
I've got no idea on the business logic but worth asking; should you include some kind of TTL for this cache?

Reply:
is not really necessary since we're processing one event at time and saving on each iteration, tbh here we just save a really few calls to the DB